pkg/operator/bootstrapteardown/bootstrap_teardown_controller.go

```diff
@@ -1,66 +1,47 @@
 package bootstrapteardown
 
 import (
-	"fmt"
+	"context"
 	"time"
 
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+
 	operatorv1 "github.com/openshift/api/operator/v1"
-	"github.com/openshift/cluster-etcd-operator/pkg/etcdcli"
-	"github.com/openshift/cluster-etcd-operator/pkg/operator/ceohelpers"
+	"github.com/openshift/library-go/pkg/controller/factory"
 	"github.com/openshift/library-go/pkg/operator/events"
 	"github.com/openshift/library-go/pkg/operator/v1helpers"
-	apierrors "k8s.io/apimachinery/pkg/api/errors"
-	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
-	"k8s.io/apimachinery/pkg/util/wait"
-	corev1listers "k8s.io/client-go/listers/core/v1"
-	"k8s.io/client-go/tools/cache"
-	"k8s.io/client-go/util/workqueue"
-	"k8s.io/klog"
-)
 
-const (
-	workQueueKey = "key"
+	"github.com/openshift/cluster-etcd-operator/pkg/etcdcli"
+	"github.com/openshift/cluster-etcd-operator/pkg/operator/ceohelpers"
 )
 
 type BootstrapTeardownController struct {
 	operatorClient  v1helpers.StaticPodOperatorClient
 	etcdClient      etcdcli.EtcdClient
 	configmapLister corev1listers.ConfigMapLister
-
-	cachesToSync  []cache.InformerSynced
-	queue         workqueue.RateLimitingInterface
-	eventRecorder events.Recorder
 }
 
 func NewBootstrapTeardownController(
 	operatorClient v1helpers.StaticPodOperatorClient,
 	kubeInformersForNamespaces v1helpers.KubeInformersForNamespaces,
-
 	etcdClient etcdcli.EtcdClient,
-
 	eventRecorder events.Recorder,
-) *BootstrapTeardownController {
+) factory.Controller {
 	c := &BootstrapTeardownController{
 		operatorClient:  operatorClient,
 		etcdClient:      etcdClient,
 		configmapLister: kubeInformersForNamespaces.InformersFor("kube-system").Core().V1().ConfigMaps().Lister(),
-
-		cachesToSync: []cache.InformerSynced{
-			operatorClient.Informer().HasSynced,
-			kubeInformersForNamespaces.InformersFor("kube-system").Core().V1().ConfigMaps().Informer().HasSynced,
-		},
-		queue:         workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "BootstrapTeardownController"),
-		eventRecorder: eventRecorder.WithComponentSuffix("bootstrap-teardown-controller"),
 	}
 
-	operatorClient.Informer().AddEventHandler(c.eventHandler())
-	kubeInformersForNamespaces.InformersFor("kube-system").Core().V1().ConfigMaps().Informer().AddEventHandler(c.eventHandler())
-
-	return c
+	return factory.New().ResyncEvery(time.Minute).WithInformers(
+		operatorClient.Informer(),
+		kubeInformersForNamespaces.InformersFor("kube-system").Core().V1().ConfigMaps().Informer(),
+	).WithSync(c.sync).ToController("BootstrapTeardownController", eventRecorder.WithComponentSuffix("bootstrap-teardown-controller"))
 }
 
-func (c *BootstrapTeardownController) sync() error {
-	err := c.removeBootstrap()
+func (c *BootstrapTeardownController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	err := c.removeBootstrap(syncCtx)
```

Comment:
Out of curiosity, why not use the .WithSyncDegraded() function of the builder, remove all the sync methods and just call the underlying action function like `removeBootstrap`? If will be more work, can be a follow-up PR..may be should add a TODO? In general IMHO it will be less code maintenance for the project

Reply:
i would take that as follow up... changing the actual logic was not the intent of this PR.